### ISSUE TITLE
Missing elliptic curves

### ISSUE THREAD

**Issue by tarcieri** [RustCrypto/elliptic-curves#114]
This is a tracking issue for additional elliptic curves we could potentially implement. Note that we are presently focusing on "legacy" short Weierstrass curves. There are many other Rust projects implementing newer curve forms, such as [curve25519-dalek](https://github.com/dalek-cryptography/curve25519-dalek).

[ ] NIST P-521
[ ] Brainpool P-256 (brainpoolP256r1)
[ ] Brainpool P-384 (brainpoolP384r1)

**Comment by tarcieri** [comment]
I added an initial set of Brainpool curves (`bp256` and `bp384`) in #281 and #280 respectively.

Unfortunately I'm not a Brainpool user, so I was unclear around usage of brainpoolP256r1/brainpoolP384r1 vs brainpoolP256t1/brainpoolP384t1. The crates I added assume the former.

Happy to hear from Brainpool users about how the curves are used in practice. If it turns out `*r1` is a mistake we can potentially transition over to `*t1` or support both `*r1` and `*t1`.

**Comment by tarcieri** [comment]
Circling back on Brainpool, I see a couple ways forward. We should decide which one makes sense.

First, for background: the `r` curves are "verifiably random" (or rather, were [supposed to be but were erroneously generated](https://bada55.cr.yp.to/brainpool.html)). The `t` curves are the "twist" curves and provide the [same effective security properties but permit a slightly more efficient implementation due to the use of smaller parameters](https://crypto.stackexchange.com/questions/42427/what-is-the-difference-between-regular-and-twisted-ecc-curves).

So, it seems like we should probably support both `r` and `t`, for both the 256-bit and 384-bit modulus sizes. Here are a couple of ways that could be done:

## Include both `r` and `t` curves in the same crate

This would result in the following crates:

- `bp256`: brainpoolP256r1 and brainpoolP256t1
- `bp384`: brainpoolP384r1 and brainpoolP384t1

## Separate crates for `r` and `t` curves

This would look like:

- `bp256r1`: brainpoolP256r1
- `bp256t1`: brainpoolP256t1
- `bp384r1`: brainpoolP384r1
- `bp384r1`: brainpoolP384t1

My vote would be for `bp256` and `bp384` crates for now, each containing both the `r` and `t` variants respectively.

**Comment by kamulos** [comment]
I know it is not one of the most widespread curves, but is there any chance, that secp224r1 is supported at some point?

**Comment by tarcieri** [comment]
@kamulos it might be possible to support sometime after #218 lands

**Comment by tarcieri** [comment]
I assume you're specifically interested in Ed25519 and RFC 8410 support?

There's an open PR to add PKCS#8 and SPKI support to `ed25519-zebra`:

https://github.com/ZcashFoundation/ed25519-zebra/pulls

I'll also note that similar functionality is available through the `signatory` crate:

https://docs.rs/signatory/0.23.2/signatory/ed25519/struct.SigningKey.html#method.from_pkcs8_der

We do also maintain the `ed25519` crate which is used by `ed25519-dalek`. I could potentially add support for decoding/encoding PKCS#8 and SPKI public keys and private keys to bytes:

https://github.com/RustCrypto/signatures/tree/master/ed25519

**Comment by tarcieri** [comment]
Now that `p384` is mostly good to go and based on what are hopefully reusable wrappers around fiat-crypto (#586), it might be possible to wrap their implementation of P-521's fields, although first I'll probably take a look at migrating the field implementations in `p256` to also use fiat-crypto.

**Comment by kamulos** [comment]
Sounds reasonable :+1: 

I need the curve for my company's product. Having this without importing one of the few C crypto libraries supporting it would be nice to have, but obviously I am aware that this is a quite exotic curve.

**Comment by tarcieri** [comment]
#607 added the initial boilerplate for a `p521` crate.

However, implementing it properly will require some breaking changes to the `elliptic-curve` crate as various properties which can presently be computed/inferred, such as the size of a serialized field element, will require explicit configuration with `p521`.

**Comment by tarcieri** [comment]
#631 extracts a proper generic implementation of `AffinePoint` and `ProjectivePoint`, which eliminates a lot of code duplication and will make it much easier to implement `p521` as well as adding arithmetic backends to `bp256` and `bp384`.

**Comment by tarcieri** [comment]
@newpavlov we got a request on Zulip for a `p224` crate. Can you grant ownership to the `elliptic-curves` team?

**Comment by newpavlov** [comment]
@tarcieri Done.

**Comment by tarcieri** [comment]
@newpavlov thank you!

**Comment by tarcieri** [comment]
`p224` and `p521` now have base and scalar field implementations, as well as curve parameter definitions, although none of it is properly tested and all gated under the `wip-arithmetic-do-not-use` crate feature.

**Comment by halvors** [comment]
Could we add secp192r1 to this list?

**Comment by tarcieri** [comment]
@halvors added, although given its small modulus size it will probably be a fairly low priority (i.e. we should finish up P-224, P-521, and the Brainpool curves first, at least)

**Comment by newpavlov** [comment]
@tarcieri 
I have added @RustCrypto/elliptic-curves to owners of `sm2` (though I don't remember exactly how I got it).

P.S.: I think it would be better to use the Zulip private chat for such questions, since I doubt other users are interested in these matters.

**Comment by halvors** [comment]
I see that `secp256k1` is supported, and seems to be the most complete so far, is this generalized? Would it be easy to support other k1 curves, for example `secp224k1` ?

**Comment by tarcieri** [comment]
@halvors supporting any given curve is a lot of work, as we need to synthesize field implementations (both base and scalar) for each curve, and plug in formulas for the parts of the field implementation which can't be synthesized

**Comment by tarcieri** [comment]
As of v0.13.1 (#820), `p224` has an `arithmetic` feature as well as support for ECDH and ECDSA

**Comment by halvors** [comment]
> @halvors supporting any given curve is a lot of work, as we need to synthesize field implementations (both base and scalar) for each curve, and plug in formulas for the parts of the field implementation which can't be synthesized.
> 
> The curve arithmetic implementation is generic; the field arithmetic implementations are not.

I fully understand that it is a lot of work, but would it still be possible to add it to the list to eventually be added in the future when the other curves are finished?

Also I can to the best of my ability help out, I don't think I am skilled enough to write the field arithmetic implementation myself, but still I can help out with other tedious work like adding test vectors etc that maybe isn't a lot of help but hopefully still can reduce the amount of work that has to be done.

**Comment by tarcieri** [comment]
Okay, I can add it, although I'd prefer to implement curves people are actually using

**Comment by halvors** [comment]
> Okay, I can add it, although I'd prefer to implement curves people are actually using

Very understandable, in my use case i have to verify signature using P-192, P-224, and k224 (but not signing) this is due to the fact that i am verifying authentication tokens issued by an external provider (somewhat dated one of that).
So me personally have a use case for it, but i understand that the curve is rather exotic and not much used in general.

Has there been any progress on generalizing k256? If so could generic implementations be reused for k224? Or is something that is yet to be done?

**Comment by fjarri** [comment]
> Has there been any progress on generalizing k256? If so could generic implementations be reused for k224? Or is something that is yet to be done?

@halvors https://github.com/RustCrypto/elliptic-curves/tree/master/primeorder seems to be it? Have you tried applying it to your curves?

**Comment by tarcieri** [comment]
Yeah, `primeorder` should work fine for it, but there are other curves I'd like to work on first which are niche but more commonly used, like the Brainpool curves

**Comment by halvors** [comment]
> Yeah, `primeorder` should work fine for it, but there are other curves I'd like to work on first which are niche but more commonly used, like the Brainpool curves

For future reference the problem with `primeorder` is that it doesn't support `n` bigger than it's size. For `secp224k1` this is 29 bytes long not 28 bytes as other 224 sized curves. AFAIK this requires a breaking change in the `primeorder` crate.

**Comment by halvors** [comment]
@tarcieri I think `secp192r1` can be marked as done in this issue, you agree?

**Comment by tarcieri** [comment]
Yes, `secp192r1` is "complete" in that high-level private key operations are deliberately not implemented due to the small modulus size, following NIST recommendations. Public-key operations are still available.

> AFAIK this requires a breaking change in the primeorder crate.

It's deeper than that, it would be a change to the `elliptic-curve` crate: https://github.com/RustCrypto/traits/issues/1304

**Comment by Unactived** [comment]
> `p224` and `p521` now have base and scalar field implementations, as well as curve parameter definitions, although none of it is properly tested and all gated under the `wip-arithmetic-do-not-use` crate feature.
https://github.com/RustCrypto/elliptic-curves/issues/114#issuecomment-1472721551

Hello, does this mean basic field arithmetic is currently expected to "work" with these curves (but untested and potentially having bugs and security flaws)?

Is it expected that this snippet features unmarked multiplication overflows (which panics in debug)?

```rust
use elliptic_curve::SecretKey;
use p521::NistP521; // with feature wip-arithmetic-do-not-use

fn main() {
    let mut rng = rand::thread_rng();

    let secret_key : SecretKey<NistP521> = SecretKey::random(&mut rng);
    let public_key = secret_key.public_key();

    println!("{:?}", public_key);
}
```

Which apparently occurs in fiat generated code, at https://github.com/RustCrypto/elliptic-curves/blob/master/p521/src/arithmetic/field/p521_64.rs#L324

Moreover if one forgets overflows (e.g. compiles in release), the snippet still does not work as multiplication here relies on inversion, which is unimplemented at the moment. So it appears a scalar multiplication cannot be performed.

I am not too familiar with the structure of the repository and its implementation ; what is the current state of the implementation of these curves, besides the lack of tests and security confidence?

Sorry if this should be in another issue, I don't know as I don't know how much this is expected.

**Comment by tarcieri** [comment]
`p521` is not expected to work, no. I'm not sure why you expected that from a feature called `wip-arithmetic-do-not-use` to indicate it's expected to work.

`p224` has since been promoted to a proper `arithmetic` feature after tests were added, but `p521` has not and has known issues.

It uses a different fiat-crypto code generator specific to Solinas primes, which is a unique property of the prime modulus of the P-521 base field. However, something is wrong with the implementation and I haven't figured out what.

I've been hoping the new codegen in `fiat-crypto` which uses newtypes in place of type aliases might help track down the issue, but haven't had time to try to update the generated code (and note in general that's a relatively difficult thing to do).

**Comment by MasterAwesome** [comment]
> I've been hoping the new codegen in `fiat-crypto` which uses newtypes in place of type aliases might help track down the issue, but haven't had time to try to update the generated code (and note in general that's a relatively difficult thing to do).
> 

I can help in creating a PR to update the existing code into the new type codegen by fiat-crypto. Would that be helpful?

**Comment by MasterAwesome** [comment]
> > `p224` and `p521` now have base and scalar field implementations, as well as curve parameter definitions, although none of it is properly tested and all gated under the `wip-arithmetic-do-not-use` crate feature.
> https://github.com/RustCrypto/elliptic-curves/issues/114#issuecomment-1427237213
> 
> Hello, does this mean basic field arithmetic is currently expected to "work" with these curves (but untested and potentially having bugs and security flaws)?
> 
> Is it expected that this snippet features unmarked multiplication overflows (which panics in debug)?
> 
> ```rust
> use elliptic_curve::SecretKey;
> use p521::NistP521; // with feature wip-arithmetic-do-not-use
> 
> fn main() {
>     let mut rng = rand::thread_rng();
> 
>     let secret_key : SecretKey<NistP521> = SecretKey::random(&mut rng);
>     let public_key = secret_key.public_key();
> 
>     println!("{:?}", public_key);
> }
> ```
> 
> Which apparently occurs in fiat generated code, at https://github.com/RustCrypto/elliptic-curves/blob/master/p521/src/arithmetic/field/p521_64.rs#L324
> 
> Moreover if one forgets overflows (e.g. compiles in release), the snippet still does not work as multiplication here relies on inversion, which is unimplemented at the moment. So it appears a scalar multiplication cannot be performed.
> 
> I am not too familiar with the structure of the repository and its implementation ; what is the current state of the implementation of these curves, besides the lack of tests and security confidence?
> 
> Sorry if this should be in another issue, I don't know as I don't know how much this is expected.

This note returns a valid public key :)

**Comment by tarcieri** [comment]
Yep, `p521` looks close now. I'm filling in some additional tests.

**Comment by drcapybara** [comment]
Any strong feelings on [Curve448-Goldilocks](https://en.wikipedia.org/wiki/Curve448)?

Seeing it [standardized](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf) and favored for inclusion in TLS might be worth the effort to investigate on our end.

**Comment by tarcieri** [comment]
@drcapybara there's already a few implementations of it out there, e.g. https://crates.io/crates/ed448-goldilocks

We could potentially ask about upstreaming them

**Comment by drcapybara** [comment]
> @drcapybara there's already a few implementations of it out there, e.g. https://crates.io/crates/ed448-goldilocks
> 
> We could potentially ask about upstreaming them

This is a solid example. It appears there may be some good opportunity to upgrade some of the [scalar arithmetic](https://github.com/crate-crypto/Ed448-Goldilocks/blob/master/src/field/scalar.rs#L13) with functionality offered by crypto-bigint here.

**Comment by samlaf** [comment]
Any plans to support bn254?

**Comment by tarcieri** [comment]
Arkworks already has an implementation of BN254: https://github.com/arkworks-rs/curves/tree/master/bn254

**Comment by halvors** [comment]
Is the required refactoring done in primeorder in order to support secp224k1 with N being 1 bit larger than the curve?
If not what is blocking this and when can it be done?

**Comment by tarcieri** [comment]
The breaking change needs to happen in the `elliptic-curve` crate, not just `primeorder`, and it probably needs some design work, i.e. new type names, which is something that's really needed revisiting for awhile. Regardless, the relevant issue and place for discussion is here: https://github.com/RustCrypto/traits/issues/1304

**Comment by AbeerHaroon** [comment]
Thank you for all of your work, it's a great endeavor. Did you get to spend any time on the Brainpool curves down the road?

**Comment by tarcieri** [comment]
Yeah, still haven't gotten to the bottom of what the issue is with them, but they should be close to having arithmetic backends

**Comment by tarcieri** [comment]
Yes, along with all of the other NIST P-curves: https://github.com/RustCrypto/elliptic-curves/tree/master/p384

**Comment by drcapybara** [comment]
> @drcapybara there's already a few implementations of it out there, e.g. https://crates.io/crates/ed448-goldilocks
> 
> We could potentially ask about upstreaming them

where does this stand currently? The last published version of this crate came in about a year ago it looks like

**Comment by tarcieri** [comment]
We could potentially inquire on the upstream repo: https://github.com/crate-crypto/Ed448-Goldilocks

**Comment by drcapybara** [comment]
> We could potentially inquire on the upstream repo: https://github.com/crate-crypto/Ed448-Goldilocks

sounds like the authors might be interested: https://github.com/crate-crypto/Ed448-Goldilocks/issues/35

**Comment by kayabaNerve** [comment]
secq256k1? Should be trivial given the secp256k1 code present.

**Comment by Fethbita** [comment]
Hi, I haven't seen `brainpoolP224r1`, `brainpoolP320r1` and `brainpoolP512r1` mentioned here. Specifically, I need the `brainpoolP512r1` as it is already used in the wild. See CSCA Master List issued by BSI: https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/ElekAusweise/CSCA/GermanMasterList.html

The Master List Signer Certificate and the German CSCA certificate uses `brainpoolP512r1`.

**Comment by dignifiedquire** [comment]
It looks like there is a newly active for for ed448, https://github.com/mikelodder7/Ed448-Goldilocks

**Comment by kayabaNerve** [comment]
I did an implementation of ed448 I'm fine transferring if that's of interest. It uses the ff/group APIs and crypto-bigint internally.

I only comment now due to the apparent interest on the subject by the recent comment.

I didn't implement Decaf however, nor anything past the curve itself.

**Comment by ashiskumarnaik** [comment]
Hi team
I’m very interested in contributing to the Weierstrass elliptic curves with in the RustCrypto project. I noticed that some of these curves are yet to be implemented (like Brainpool P-512). @tarcieri  @newpavlov can you point me to any specific one?, would love to discuss further!

**Comment by AaronKutch** [comment]
Where is the new implementation of X448? If I understand correctly, `ed448_goldilocks` is just the library that provides low level point definitions, and there needs to be one more layer. The repo I was using just got set into read-only mode (https://github.com/crate-crypto/x448/pull/7) so I can't ask to find out where the new one is.

**Comment by tarcieri** [comment]
@AaronKutch #1250
